pkgs/by-name/cr/criterion/package.nix

```diff
@@ -0,0 +1,137 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  boxfort,
+  meson,
+  libcsptr,
+  pkg-config,
+  gettext,
+  cmake,
+  ninja,
+  protobuf,
+  libffi,
+  libgit2,
+  dyncall,
+  nanomsg,
+  nanopbMalloc,
+  python3Packages,
+  testers,
+  criterion,
+}:
+
+let
+  # follow revisions defined in .wrap files
+  debugbreak = fetchFromGitHub {
+    owner = "MrAnno";
+    repo = "debugbreak";
+    rev = "83bf7e933311b88613cbaadeced9c2e2c811054a";
+    sha256 = "sha256-OPrPGBUZN73Nl5NMEf/nME843yTolt913yjut3rAos0=";
+  };
+
+  klib = fetchFromGitHub {
+    owner = "attractivechaos";
+    repo = "klib";
+    rev = "cdb7e9236dc47abf8da7ebd702cc6f7f21f0c502";
+    sha256 = "sha256-+GaI5nXz4jYI0rO17xDhNtFpLlGL2WzeSVLMfB6Cl6E=";
+  };
+in
+stdenv.mkDerivation rec {
+  pname = "criterion";
+  version = "2.4.2";
+
+  src = fetchFromGitHub {
+    owner = "Snaipe";
+    repo = "Criterion";
+    rev = "v${version}";
+    sha256 = "sha256-5GH7AYjrnBnqiSmp28BoaM1Xmy8sPs1atfqJkGy3Yf0=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [
+    meson
+    ninja
+    cmake
+    pkg-config
+    protobuf
+  ];
+
+  buildInputs = [
+    boxfort.dev
+    dyncall
+    gettext
+    libcsptr
+    nanomsg
+    nanopbMalloc
+    libgit2
+    libffi
+  ];
+
+  nativeCheckInputs = with python3Packages; [ cram ];
+
+  doCheck = true;
+
+  prePatch = ''
+    cp -r ${debugbreak} subprojects/debugbreak
+    cp -r ${klib} subprojects/klib
+
+    for dep in "debugbreak" "klib"; do
+      local meson="$dep/meson.build"
+
+      chmod +w subprojects/$dep
+      cp subprojects/packagefiles/$meson subprojects/$meson
+    done
+  '';
+
+  postPatch = ''
+    patchShebangs ci/isdir.py src/protocol/gen-pb.py
+  '';
+
+  outputs = [
+    "dev"
+    "out"
+  ];
+
+  passthru.tests = {
+    version =
```

Comment:
1. A full custom test is better suited here.
2. It's better to transplant this expression in a proper file, say `tests/001-version.nix`.

Reply:
> 1. A full custom test is better suited here.

What do you mean by a full suited test?
I made a nixos test before, but i was requested to remove it
- see: https://github.com/NixOS/nixpkgs/pull/308951#discussion_r1589924732

## 

> 2. It's better to transplant this expression in a proper file, say `tests/001-version.nix`.

Other derivation doesn't seem to follow this logic:
- [packages with `passthru.tests.version`](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20passthru.tests.version&type=code)

Reply:
> 1. A full custom test is better suited here.

@AndersonTorres Please clarify what you mean by a "full custom test". Do you mean a NixOS test? I wouldn't do that since Criterion does not offer a NixOS service or component otherwise related to NixOS specifically.

> > 2. It's better to transplant this expression in a proper file, say `tests/001-version.nix`
>
> Other derivation doesn't seem to follow this logic:
>
> * [packages with passthru.tests.version](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20passthru.tests.version&type=code)

@Sigmanificient I don't think it invalidates the point, better late than never to apply a good practice.

Reply:
>     * see: [criterion: 2.4.1 -> 2.4.2, add maintainer, migrate by-name, add version check #308951 (comment)](https://github.com/NixOS/nixpkgs/pull/308951#discussion_r1589924732)

Indeed a NixOS test is too heavy to test a package. NixOS tests are better suited for testing NixOS modules.

> Other derivation doesn't seem to follow this logic:
> 
>     * [packages with `passthru.tests.version`](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs%20passthru.tests.version&type=code)

Many of the examples cited are using templates, especially version tests. Since many programs have a flag to emit their version, it is natural to create a template for them.
It does not mean that custom testing scripts/expressions can't or shouldn't be created.

Think in e.g. update scripts. There are many templates for update scripts, and indeed they cover a lot of the Nixpkgs packages.
Nonetheless, there are also many custom scripts (like the binary-only releases for Brave browser).

Since you are creating a package to execute a test, it is natural to transplant this package to a separate file.

> Please clarify what you mean by a "full custom test". Do you mean writing a lambda function to generate test derivations?

This is not needed. Transplanting the test derivation to a separate file and calling it via `callPackage` suffices.
An approximated example can be seen at elvish expression:
https://github.com/NixOS/nixpkgs/tree/9672bdfc29e4a275c4230f0aaf51bb5e1c907b1f/pkgs/by-name/el/elvish